pkg/xds/translate.go

```diff
@@ -0,0 +1,287 @@
+type ClusterName func(*routingv2.Destination) string
+
+// TranslateHost creates virtual host routes from the v1alpha2 config.
+func TranslateHost(in model.Config, name ClusterName, defaultCluster string, labels model.Labels) []route.Route {
+	rule, ok := in.Spec.(*routingv2.RouteRule)
+	if !ok {
+		return nil
+	}
+
+	operation := in.ConfigMeta.Name
+
+	if len(rule.Http) == 0 {
+		return []route.Route{
+			TranslateRoute(&routingv2.HTTPRoute{}, nil, operation, name, defaultCluster),
+		}
+	}
+
+	out := make([]route.Route, 0)
+	for _, http := range rule.Http {
+		if len(http.Match) == 0 {
+			out = append(out, TranslateRoute(http, nil, operation, name, defaultCluster))
+		} else {
+			for _, match := range http.Match {
+				if model.Labels(match.SourceLabels).SubsetOf(labels) {
+					out = append(out, TranslateRoute(http, match, operation, name, defaultCluster))
+				}
+			}
+		}
+	}
+
+	return out
+}
+
+// TranslateRoute translates HTTP routes
+// TODO: fault filters -- issue https://github.com/istio/api/issues/388
+func TranslateRoute(in *routingv2.HTTPRoute,
+	match *routingv2.HTTPMatchRequest,
+	operation string,
+	name ClusterName,
+	defaultCluster string) route.Route {
+	out := route.Route{
+		Match: TranslateRouteMatch(match),
+		Decorator: &route.Decorator{
+			Operation: operation,
+		},
+	}
+
+	if redirect := in.Redirect; redirect != nil {
+		out.Action = &route.Route_Redirect{
+			Redirect: &route.RedirectAction{
+				HostRedirect: redirect.Authority,
+				PathRedirect: redirect.Uri,
+			}}
+	} else {
+		action := &route.RouteAction{
+			Cors:         TranslateCORSPolicy(in.CorsPolicy),
+			RetryPolicy:  TranslateRetryPolicy(in.Retries),
+			Timeout:      TranslateTime(in.Timeout),
+			UseWebsocket: &types.BoolValue{Value: in.WebsocketUpgrade},
+		}
+		out.Action = &route.Route_Route{Route: action}
+
+		if rewrite := in.Rewrite; rewrite != nil {
+			action.PrefixRewrite = rewrite.Uri
+			action.HostRewriteSpecifier = &route.RouteAction_HostRewrite{
+				HostRewrite: rewrite.Authority,
+			}
+		}
+
+		if len(in.AppendHeaders) > 0 {
+			action.RequestHeadersToAdd = make([]*core.HeaderValueOption, len(in.AppendHeaders))
+			for key, value := range in.AppendHeaders {
+				action.RequestHeadersToAdd = append(action.RequestHeadersToAdd, &core.HeaderValueOption{
+					Header: &core.HeaderValue{
+						Key:   key,
+						Value: value,
+					},
+				})
+			}
+		}
+
+		if in.Mirror != nil {
+			action.RequestMirrorPolicy = &route.RouteAction_RequestMirrorPolicy{Cluster: name(in.Mirror)}
+		}
+
+		if len(in.Route) == 0 { // build default cluster
+			action.ClusterSpecifier = &route.RouteAction_Cluster{Cluster: defaultCluster}
+		} else {
+			weighted := make([]*route.WeightedCluster_ClusterWeight, len(in.Route))
+			for _, dst := range in.Route {
+				weighted = append(weighted, &route.WeightedCluster_ClusterWeight{
+					Name:   name(dst.Destination),
+					Weight: &types.UInt32Value{Value: uint32(dst.Weight)},
```

Comment:
can you add the label metadata as metadata in this unit? This will allow us to turn on subset load balancing as an option in future.

Reply:
@costinm wants to keep cluster names the same, so switching to subset LB is neither an option nor a blocker for this cycle.